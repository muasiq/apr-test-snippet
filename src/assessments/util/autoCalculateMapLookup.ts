export enum AutoCalculateMapConfigOption {
	DEFAULT = 'Default',
	CHOICE = 'Choice',
}
