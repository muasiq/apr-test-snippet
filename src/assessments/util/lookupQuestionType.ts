import { AssessmentQuestionType, QuestionBase } from '../types';

export function lookupQuestionType(q: QuestionBase): AssessmentQuestionType {
	if (['select', 'single select', 'code'].includes(q.questionType?.toLowerCase() ?? '')) {
		return AssessmentQuestionType.SelectOne;
	}
	if (['multiselect', 'checklist', 'select all that apply'].includes(q.questionType?.toLowerCase() ?? '')) {
		return AssessmentQuestionType.SelectAllThatApply;
	}
	if (q.questionType?.toLowerCase() === 'text') {
		return AssessmentQuestionType.FreeForm;
	}
	if (q.questionType?.toLowerCase() === 'date') {
		return AssessmentQuestionType.Date;
	}
	if (['number', 'numeric'].includes(q.questionType?.toLowerCase() ?? '')) {
		return AssessmentQuestionType.FreeFormNumber;
	}
	if (q.questionType?.toLowerCase() === 'boolean') {
		return AssessmentQuestionType.SelectOne;
	}
	if (!q.questionType && q.type?.toLowerCase() === 'panel') {
		return AssessmentQuestionType.Panel;
	}
	if (q.questionType?.toLowerCase() === 'longtext') {
		return AssessmentQuestionType.LongText;
	}
	if (q.rowType?.toLowerCase() === 'section') {
		return AssessmentQuestionType.Panel;
	}

	throw new Error(`Unknown question type: ${q.questionType}`);
}
