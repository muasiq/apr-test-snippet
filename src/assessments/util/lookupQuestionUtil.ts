/* eslint-disable @typescript-eslint/naming-convention */
import { AssessmentAnswer } from '@prisma/client';
import { get } from 'lodash';
import { addMultipleIndex } from '../../common/utils/addMultipleIndex';
import { resolveIdAndMultipleIndex } from '../../common/utils/resolveIdAndMultipleIndex';
import { PatientWithJoins } from '../../server/api/routers/patient/patient.types';
import { Configuration } from '../configurations';
import { PreReqMap } from '../oasis/prerequisites/PreReqMap';
import {
	AssessmentQuestionSources,
	AssessmentQuestionType,
	SubGroup,
	type AssessmentQuestion,
	type Group,
	type QuestionSource,
} from '../types';

export function lookupQuestion(questionId: string, configuration: Configuration) {
	const { resolvedQuestionId, index } = resolveIdAndMultipleIndex(questionId);
	const found = configuration.allQuestions.find((q) => q.id === resolvedQuestionId);
	if (!found) throw new Error(`Question with id ${questionId} not found`);
	return { ...found, id: addMultipleIndex(resolvedQuestionId, index) };
}

export function lookupQuestionOrNull(questionId: string, configuration: Configuration) {
	try {
		return lookupQuestion(questionId, configuration);
	} catch (e) {
		return null;
	}
}

export const findParentQuestionFromFollowup = (questionId: string, configuration: Configuration) => {
	const { resolvedQuestionId, index } = resolveIdAndMultipleIndex(questionId);
	const found = configuration.allQuestions.find((q) => {
		return q.responses?.flatMap((r) => r.followup).includes(resolvedQuestionId) ?? false;
	});
	if (!found) {
		return;
	}
	return { ...found, id: addMultipleIndex(found.id, index) };
};

export const findTopLevelParentQuestionFromFollowup = (questionId: string, configuration: Configuration) => {
	const { resolvedQuestionId, index } = resolveIdAndMultipleIndex(questionId);

	const parentLookup: Record<string, AssessmentQuestion> = {};
	configuration.allQuestions.forEach((q) => {
		q.responses?.forEach((response) => {
			response.followup?.forEach((f) => {
				parentLookup[f] = q;
			});
		});
	});

	let current = configuration.allQuestions.find((q) => q.id === resolvedQuestionId);
	if (!current) throw new Error(`Question with id ${questionId} not found`);

	const visited = new Set<string>();
	while (parentLookup[current.id] && !visited.has(current.id)) {
		visited.add(current.id);
		current = parentLookup[current.id]!;
	}

	return { ...current, id: addMultipleIndex(current.id, index) };
};

export const getAllConfigQuestionsFlat = (configuration: Configuration, patientData: PatientWithJoins) => {
	const allQuestions = configuration.qaConfig.flatMap((g) => extractQuestionsFromGroup(g, patientData));
	return allQuestions;
};

export const getFirstLevelFollowUps = (questions: QuestionSource[], configuration: Configuration) => {
	const allTopQuestions = questions
		.map((question) => {
			if (!question.id) {
				console.log({ question });
				console.log({ questions });
			}
			return lookupQuestion(question.id!, configuration);
		})
		.filter((a) => a);
	const allFollowUps = allTopQuestions
		.flatMap((question) => {
			return question.responses?.flatMap((response) => (response.followup ?? []).flatMap((f) => ({ id: f })));
		})
		.filter((a) => a) as { id: string }[];
	return allFollowUps;
};

export const recursivelyGetAllFollowUps = (
	questions: QuestionSource[],
	configuration: Configuration,
	finalList: QuestionSource[] = [],
): QuestionSource[] => {
	if (questions.length === 0) {
		return finalList;
	} else {
		const allFollowUps = getFirstLevelFollowUps(questions, configuration);
		return recursivelyGetAllQuestionsAndFollowUps(allFollowUps, configuration, [...finalList, ...allFollowUps]);
	}
};

export const recursivelyGetAllQuestionsAndFollowUps = (
	questions: QuestionSource[],
	configuration: Configuration,
	finalList: QuestionSource[] = [],
): QuestionSource[] => {
	if (questions.length === 0) {
		return finalList;
	} else {
		const allFollowUps = getFirstLevelFollowUps(questions, configuration);
		return recursivelyGetAllQuestionsAndFollowUps(allFollowUps, configuration, [...finalList, ...questions]);
	}
};

export const extractTopLevelQuestionsThatShouldHaveSuggestionsGenerated = (
	configuration: Configuration,
	patientData: PatientWithJoins,
) => {
	const match = (question?: QuestionSource) => {
		if (!question) return false;
		if (!question.id) return false;
		const lookup = lookupQuestion(question.id, configuration);
		if (lookup.mappedType === AssessmentQuestionType.Panel) return false;
		if (question.source === AssessmentQuestionSources.Custom && lookup.suggestionGenerator) return true;
		return question.source !== AssessmentQuestionSources.Custom && !question.preRequisiteSatisfied;
	};
	return configuration.qaConfig.flatMap((g) => extractQuestionsFromGroup(g, patientData).filter(match));
};

export const extractRelevantQuestionsFromGroup = (
	group: Group | undefined,
	answers: AssessmentAnswer[],
	configuration: Configuration,
	patientData: PatientWithJoins,
): QuestionSource[] => {
	if (!group) return [];
	const relevant = (question: QuestionSource) => {
		if (!question.id) return false;
		const lookup = lookupQuestion(question.id, configuration);
		if (lookup.mappedType === AssessmentQuestionType.Panel) return false;
		if (question.preRequisiteSatisfied) {
			const lookup = PreReqMap[question.preRequisiteSatisfied];
			return lookup(answers, configuration);
		}
		return true;
	};

	const list = extractQuestionsFromGroup(group, patientData).filter(relevant);
	return list;
};

export const extractQuestionsFromGroup = (group: Group, patientData: PatientWithJoins) => {
	return group.subGroups.flatMap((sg) => {
		if (sg.multiple) {
			const repeatTimes = get(patientData, sg.multiple.repeatTimesFieldResolver, 0) as number;
			if (repeatTimes) {
				return Array.from({ length: repeatTimes }, (_, i) => i + 1).flatMap((i) => {
					return sg.multiple!.questions.map((q) => ({ ...q, id: addMultipleIndex(q.id!, i) }));
				});
			}
			return [] as QuestionSource[];
		}
		return sg.questions;
	});
};

export const extractAllQuestionsFromGroup = (group: Group, patientData: PatientWithJoins) => {
	const subgroupQuestions = extractQuestionsFromGroup(group, patientData);
	const subHeadingQuestions = group.subHeadings?.flatMap((sh) => extractQuestionsFromGroup(sh, patientData)) ?? [];
	return [...subgroupQuestions, ...subHeadingQuestions];
};

export function findAllRollupGroups(groups: Group[]): SubGroup[] {
	return groups.flatMap((g) => {
		if (g.subGroups.length) {
			const found = g.subGroups.filter((sg) => !!sg.rollupForRpaApi);
			return found;
		}
		if (g.subHeadings?.length) {
			return findAllRollupGroups(g.subHeadings);
		}
		return [];
	});
}

export const getQuestionMetadata = (
	question: AssessmentQuestion,
	configuration: Configuration,
	patientData: PatientWithJoins,
	attempts = 0,
): QuestionSource => {
	const MAX_RECURSIVE_ATTEMPTS = 5;
	const allQuestions = getAllConfigQuestionsFlat(configuration, patientData);
	const lookup = allQuestions.find((aq) => aq.id === question.id);
	if (!lookup) {
		if (attempts > MAX_RECURSIVE_ATTEMPTS) {
			console.error(
				`Could not find question metadata for ${question.id} after ${attempts} attempts, using default`,
			);
			return {} as QuestionSource;
		} else {
			const parent = findParentQuestionFromFollowup(question.id, configuration);
			if (!parent) {
				return {} as QuestionSource;
			}
			return getQuestionMetadata(parent, configuration, patientData, attempts + 1);
		}
	}
	return lookup;
};

export const getQuestionSubGroupForMultipleTypeQuestion = (
	question: AssessmentQuestion,
	configuration: Configuration,
): SubGroup | undefined => {
	return configuration.qaConfig
		.flatMap((g) => g.subGroups)
		.find((sg) => {
			return sg.multiple?.questions.find((q) => q.id === question.id);
		});
};
