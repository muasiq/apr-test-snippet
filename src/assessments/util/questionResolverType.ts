import { QuestionSource } from '../types';

export enum QuestionResolverTypes {
	ASSESSMENT = 'Assessment',
	NOTE = 'Note',
	PATIENT_DATA = 'PatientData',
	AUTO_CALCULATE = 'AutoCalculate',
}

export function getQuestionResolverType(question: QuestionSource) {
	if (question.id && question.source) {
		return QuestionResolverTypes.ASSESSMENT;
	}
	if (question.note) {
		return QuestionResolverTypes.NOTE;
	}
	if (question.patientDataResolver) {
		return QuestionResolverTypes.PATIENT_DATA;
	}
	if (question.autoCalculateFieldResolver) {
		return QuestionResolverTypes.AUTO_CALCULATE;
	}

	console.error('unknown question resolver type:', question);
	throw new Error('unknown question resolver');
}
