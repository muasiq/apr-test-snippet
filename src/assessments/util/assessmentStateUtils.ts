import { AssessmentAnswer, AssessmentAnswerSuggestionStatus } from '@prisma/client';
import { compact, difference, isEmpty } from 'lodash';
import { Configuration } from '~/assessments/configurations';
import { AssessmentQuestion } from '~/assessments/types';
import { extractRelevantQuestionsFromGroup } from '~/assessments/util/lookupQuestionUtil';
import { addMultipleIndex } from '~/common/utils/addMultipleIndex';
import { resolveIdAndMultipleIndex } from '~/common/utils/resolveIdAndMultipleIndex';
import { responseMatches } from '~/common/utils/responseMatches';
import { CheckedResponseUnion } from '~/server/api/routers/assessment/assessment.inputs';
import { PatientWithJoins } from '~/server/api/routers/patient/patient.types';

export function hasAcceptedAnswer(answer?: AssessmentAnswer) {
	return !!answer?.generatedResponseAccepted || !isEmpty(answer?.checkedResponse);
}

export function answerSuggestionStillLoading(answer?: AssessmentAnswer) {
	return (
		!answer ||
		answer.status === AssessmentAnswerSuggestionStatus.InProgress ||
		answer.status === AssessmentAnswerSuggestionStatus.Queued
	);
}

export function pageHeadingHasAllRelevantQuestionsConfirmed(
	heading: string,
	{ configuration, patient, answers }: ConfigurationPatientAndAnswers,
) {
	const allRelevantQuestions = allRelevantQuestionsForPageHeading(heading, { configuration, patient, answers });
	return allRelevantQuestions.every((q) => {
		const confirmed = questionHasConfirmedAnswer(q, answers);
		return confirmed;
	});
}

export function allRelevantQuestionsForPageHeading(
	heading: string,
	{ configuration, patient, answers }: ConfigurationPatientAndAnswers,
) {
	const matchedGroup = configuration.qaConfig.find((g) => g.heading === heading);
	if (!matchedGroup) throw new Error(`Group with heading ${heading} not found`);
	const topLevelQuestionsInThisGroup = extractRelevantQuestionsFromGroup(
		matchedGroup,
		answers,
		configuration,
		patient,
	);
	const topLevelQuestionIdsInThisGroup = compact(topLevelQuestionsInThisGroup.map((q) => q.id));
	const allQuestions = allQuestionsMap(configuration);
	const allFollowUps = recursivelyGetAllFollowUpsForTopLevelQuestionIds(
		allQuestions,
		topLevelQuestionIdsInThisGroup,
		{ configuration, patient, answers },
	);
	return compact([...topLevelQuestionIdsInThisGroup, ...allFollowUps]);
}

export function recursivelyGetAllFollowUpsForTopLevelQuestionIds(
	allQuestions: AllQuestionsMap,
	topLevelQuestionIds: string[],
	{ configuration, patient, answers }: ConfigurationPatientAndAnswers,
	finalList: string[] = [],
) {
	if (topLevelQuestionIds.length === 0) return finalList;
	const relevantAnswers = answers.filter((a) => topLevelQuestionIds.includes(a.assessmentNumber));
	const allFollowUps = relevantAnswers.flatMap((a) => lookupFollowUpsForAnswer(a, allQuestions));
	return recursivelyGetAllFollowUpsForTopLevelQuestionIds(
		allQuestions,
		allFollowUps,
		{ configuration, patient, answers },
		[...finalList, ...allFollowUps],
	);
}

export function allRelevantQuestionIdsForPatient({ configuration, patient, answers }: ConfigurationPatientAndAnswers) {
	const allRelevantTopLevelQuestions = allTopLevelQuestionsForConfigurationAndPatient({
		configuration,
		patient,
		answers,
	});
	const allRelevantTopLevelQuestionIds = compact(allRelevantTopLevelQuestions.map((q) => q.id));
	const allQuestions = allQuestionsMap(configuration);
	const allRelevantFollowUps = recursivelyGetAllFollowUpsForTopLevelQuestionIds(
		allQuestions,
		allRelevantTopLevelQuestionIds,
		{ configuration, patient, answers },
	);
	return compact([...allRelevantTopLevelQuestionIds, ...allRelevantFollowUps]);
}

export function allRelevantQuestionsForPatientHaveConfirmedAnswers({
	configuration,
	patient,
	answers,
}: ConfigurationPatientAndAnswers) {
	const allRelevantQuestionIds = allRelevantQuestionIdsForPatient({ answers, configuration, patient });
	return allRelevantQuestionIds.every((q) => questionHasConfirmedAnswer(q, answers));
}

export function questionsThatDoNotHaveConfirmedAnswers({
	configuration,
	patient,
	answers,
}: ConfigurationPatientAndAnswers) {
	const allRelevantQuestionIds = allRelevantQuestionIdsForPatient({ answers, configuration, patient });
	const allConfirmed = allConfirmedQuestionsForPatient({ configuration, patient, answers });
	return difference(allRelevantQuestionIds, allConfirmed);
}

export function allTopLevelQuestionsForConfigurationAndPatient({
	configuration,
	patient,
	answers,
}: ConfigurationPatientAndAnswers) {
	const allQuestions = configuration.qaConfig.flatMap((g) =>
		extractRelevantQuestionsFromGroup(g, answers, configuration, patient),
	);
	return allQuestions;
}

export type AllQuestionsMap = Map<string, AssessmentQuestion>;

function lookupFollowUpsForAnswer(answer: AssessmentAnswer, allQuestions: AllQuestionsMap) {
	const { index, resolvedQuestionId } = resolveIdAndMultipleIndex(answer.assessmentNumber);
	const currentChoice = answer.checkedResponse ?? answer.generatedResponse;
	const match = allQuestions.get(resolvedQuestionId);
	if (!match) return [];
	const followUps = match.responses?.flatMap((r) => r.followup) ?? [];
	if (!followUps.length) return [];
	const matchingResponses = match.responses?.filter((r) =>
		responseMatches((currentChoice as CheckedResponseUnion)?.choice as string, r.text),
	);
	if (matchingResponses?.length) {
		if (index) {
			return matchingResponses.flatMap((r) => r.followup?.map((f) => addMultipleIndex(f, index)) ?? []);
		}
		return matchingResponses.flatMap((r) => r.followup ?? []);
	}
	return [];
}

type ConfigurationPatientAndAnswers = {
	configuration: Configuration;
	patient: PatientWithJoins;
	answers: AssessmentAnswer[];
};

export function allQuestionsMap(configuration: Configuration): AllQuestionsMap {
	const map = new Map(configuration.allQuestions.map((q) => [q.id, q]));
	return map;
}

function allConfirmedQuestionsForPatient({ configuration, patient, answers }: ConfigurationPatientAndAnswers) {
	const allRelevantQuestionIds = allRelevantQuestionIdsForPatient({ answers, configuration, patient });
	return allRelevantQuestionIds.filter((q) => questionHasConfirmedAnswer(q, answers));
}

function questionHasConfirmedAnswer(questionId: string, answers: AssessmentAnswer[]) {
	const answer = answers.find((a) => a.assessmentNumber === questionId);
	if (!answer) return false;
	const confirmed = hasAcceptedAnswer(answer);
	return confirmed;
}
