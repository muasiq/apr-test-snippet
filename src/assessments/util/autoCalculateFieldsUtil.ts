import { RouterOutputs } from '~/common/utils/api';
import { AdditionalDropdownConfigurationInput } from '~/server/api/routers/organization/organization.inputs';
import { Configuration } from '../configurations';
import { lookupQuestion } from './lookupQuestionUtil';

interface CheckedResponse {
	checkedResponse: {
		choice: string;
	};
}

export type AutoCalculateFieldResolver = {
	assessmentData: (RouterOutputs['assessment']['getAllAssessmentQuestionsForPatient'][number] & CheckedResponse)[];
	configuration: Configuration;
	additionalDropdownConfigurationInput: AdditionalDropdownConfigurationInput;
	readOnly: boolean;
	print?: boolean;
};

export const getUnansweredQuestions = <T extends Record<string, string>>(
	assessmentData: AutoCalculateFieldResolver['assessmentData'],
	questionEnum: T,
) => {
	return Object.values(questionEnum).filter(
		(question) => !assessmentData.some((item) => item.assessmentNumber === question && item.checkedResponse),
	);
};

export const getAnsweredAssessmentQuestionsByEnum = <
	T extends Record<string, string>,
	U extends Record<'assessmentNumber', string> & { checkedResponse: unknown },
>(
	assessmentData: U[],
	questionEnum: T,
) => {
	const questionEnumSet = new Set(Object.values(questionEnum));
	return assessmentData.filter((item) => questionEnumSet.has(item.assessmentNumber) && item.checkedResponse);
};

export const getDataFileChoiceScore = (
	item: Partial<AutoCalculateFieldResolver['assessmentData'][0]> | undefined,
	configuration: Configuration | null,
): number => {
	if (!configuration) return 0;
	if (!item?.checkedResponse || !item.assessmentNumber) return 0;

	const question = lookupQuestion(item.assessmentNumber, configuration);

	if (!question?.responses) {
		console.error(`${item.assessmentNumber} question or responses not found`);
		return 0;
	}

	const response = question.responses.find(
		(r) => r.text?.trim()?.toLowerCase() === item.checkedResponse?.choice?.trim().toLowerCase(),
	);

	if (!response?.text) {
		console.error(`${item.assessmentNumber} response not found for choice:`, item.checkedResponse.choice);
		return 0;
	}

	if (response.code) {
		return parseInt(response.code);
	}

	const position = question.responses.indexOf(response);
	if (position < 0) {
		console.error(`${item.assessmentNumber} response not found for choice:`, item.checkedResponse.choice);
		return 0;
	}
	return position + 1;
};
