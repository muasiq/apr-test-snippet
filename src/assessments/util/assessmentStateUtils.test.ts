import { expect } from '@jest/globals';
import { AssessmentAnswer } from '@prisma/client';
import { compact, intersection } from 'lodash';
import { CustomAssessmentNumber } from '~/assessments/apricot/custom';
import { WOUND_HEADING } from '~/assessments/apricot/util';
import { Configuration, getLocalConfiguration } from '~/assessments/configurations';
import {
	allQuestionsMap,
	allRelevantQuestionIdsForPatient,
	allRelevantQuestionsForPageHeading,
	allRelevantQuestionsForPatientHaveConfirmedAnswers,
	allTopLevelQuestionsForConfigurationAndPatient,
	pageHeadingHasAllRelevantQuestionsConfirmed,
	questionsThatDoNotHaveConfirmedAnswers,
	recursivelyGetAllFollowUpsForTopLevelQuestionIds,
} from '~/assessments/util/assessmentStateUtils';
import { addMultipleIndex } from '~/common/utils/addMultipleIndex';
import { PatientWithJoins } from '~/server/api/routers/patient/patient.types';

describe('assessmentStateUtils', () => {
	const patient = {
		NurseInterview: {
			totalWounds: 1,
		},
	} as PatientWithJoins;

	const configuration = getLocalConfiguration('Accentra');
	const allIdsInPatientInformationSection = [
		'PA-97a6bd',
		'A1005',
		'A1010',
		'MQ29KAPBEQ',
		'A1110B',
		'PA-710511',
		'SSA102',
		'PA-329f77',
		'PA-2c33c4',
		'B1300',
		'PA-5a8f12',
		'M1100',
		'PA-1a4639',
		'PA-5a0310',
		'PA-401818',
		'PA-7241b1',
		'SSA100',
		'M2102F',
		'PA-59767e',
		'PA-bcf065',
		'PA-317452',
		'A1250',
		'PA-a86932',
		'PA-1386f9',
		'PA-c02a64',
		'advance_directives_custom',
		'facilities_custom',
		'EV00',
	];

	describe(allTopLevelQuestionsForConfigurationAndPatient.name, () => {
		const questions = allTopLevelQuestionsForConfigurationAndPatient({ configuration, patient, answers: [] });

		it('should return all top level questions for patient', () => {
			const itemsToSpotCheck = [
				'MQ373KM2QO',
				CustomAssessmentNumber.HOME_HEALTH_AIDE_VISIT_FREQUENCIES,
				'PA-2b0c08',
				'C1310A',
			];
			const spotCheck = questions.filter((q) => {
				return itemsToSpotCheck.some((item) => item === q.id);
			});
			expect(spotCheck.length).toBe(itemsToSpotCheck.length);
		});

		it('should not return follow-ups', () => {
			const followUpId = 'AC1001';
			expect(questions.find((q) => q.id === followUpId)).toBeUndefined();
		});
	});

	describe(recursivelyGetAllFollowUpsForTopLevelQuestionIds.name, () => {
		const answers = [
			generateAnswer('PSIG25V2GVRD', 'Not Met'),
			generateAnswer('PA-fcab86'),
			generateAnswer('PA-15bc7b', 'Yes', false),
			generateAnswer('PA-9f8d09', 'Yes', false),
			generateAnswer('O0110_A', ['Oxygen therapy', 'Chemotherapy'], true),
			generateAnswer('SSA01-#1'),
		];
		const allQuestions = allQuestionsMap(configuration);
		const topLevelQuestions = allTopLevelQuestionsForConfigurationAndPatient({ configuration, patient, answers });
		const topLevelQuestionIds = compact(topLevelQuestions.map((q) => q.id));
		it('should return all relevant follow ups based on generated and confirmed answers', () => {
			const relevantFollowUps = recursivelyGetAllFollowUpsForTopLevelQuestionIds(
				allQuestions,
				topLevelQuestionIds,
				{ answers, configuration, patient },
			);
			// PSIG25V2GVRD follow-up intentionally not included since it's parent is not adding it to the follow-up tree
			expect(relevantFollowUps).toEqual([
				'PA-acc548',
				'PA-f8ace2',
				'PA-9f8d09',
				'O0110_A_A1',
				'O0110_A_C1',
				'MQ6BB554AY-#1',
				'MQ55CXJBTX-#1',
				'MQ3M1T78JB-#1',
				'MQ4LWHY55V-#1',
				'MQ1A9R6EWB-#1',
				'SSA02-#1',
				'SSA04-#1',
				'PA-772181',
			]);
		});
	});

	describe(allRelevantQuestionIdsForPatient.name, () => {
		it('should return all top level questions and currently relevant follows ups based on generated and accepted answers', () => {
			const answers = [generateAnswer('PA-15bc7b'), generateAnswer('PA-fcab86'), generateAnswer('SSA01-#2')];
			const relevantQuestions = allRelevantQuestionIdsForPatient({ answers, configuration, patient });
			const itemsToSpotCheck = [
				'PA-9f8d09',
				'PA-acc548',
				'PA-f8ace2',
				CustomAssessmentNumber.HOME_HEALTH_AIDE_VISIT_FREQUENCIES,
				'PA-2b0c08',
				'C1310A',
			];
			const spotCheck = intersection(relevantQuestions, itemsToSpotCheck);
			expect(spotCheck.length).toEqual(itemsToSpotCheck.length);
		});
	});

	describe(allRelevantQuestionsForPageHeading.name, () => {
		const patient = {
			NurseInterview: {
				totalWounds: 2,
			},
		} as PatientWithJoins;
		it('should return all relevant questions for a page heading', () => {
			const heading = 'Patient Information';
			const questions = allRelevantQuestionsForPageHeading(heading, { configuration, patient, answers: [] });
			expect(questions).toEqual(allIdsInPatientInformationSection);
		});

		it('should return all relevant questions for a page heading with multiple index and no answers', () => {
			const questions = allRelevantQuestionsForPageHeading(WOUND_HEADING, {
				configuration,
				patient,
				answers: [],
			});
			expect(questions).toEqual(['SSA01-#1', 'SSA01-#2']);
		});

		it('should return all relevant questions for a page heading with multiple index and some answers', () => {
			const answers = [generateAnswer('SSA01-#2')];
			const questions = allRelevantQuestionsForPageHeading(WOUND_HEADING, {
				configuration,
				patient,
				answers,
			});
			expect(questions).toEqual([
				'SSA01-#1',
				'SSA01-#2',
				'MQ6BB554AY-#2',
				'MQ55CXJBTX-#2',
				'MQ3M1T78JB-#2',
				'MQ4LWHY55V-#2',
				'MQ1A9R6EWB-#2',
				'SSA02-#2',
				'SSA04-#2',
			]);
		});

		it('should return all relevant questions for multiple follow-up levels', () => {
			const heading = 'Health History';
			const answers = [generateAnswer('PA-15bc7b'), generateAnswer('PA-9f8d09')];
			const questions = allRelevantQuestionsForPageHeading(heading, { configuration, patient, answers });
			expect(questions).toMatchSnapshot();
			const spotCheck = ['PA-9f8d09', 'PA-772181'];
			const spotCheckMatches = questions.filter((q) => spotCheck.includes(q));
			expect(spotCheckMatches.length).toBe(spotCheck.length);
		});
	});

	describe(pageHeadingHasAllRelevantQuestionsConfirmed.name, () => {
		it('should return true if all relevant questions are confirmed', () => {
			const allQuestions = allQuestionsMap(configuration);
			const topLevelAnswers = allIdsInPatientInformationSection.map((id) => generateAnswer(id, 'No'));
			const followUps = recursivelyGetAllFollowUpsForTopLevelQuestionIds(
				allQuestions,
				allIdsInPatientInformationSection,
				{ answers: topLevelAnswers, configuration, patient },
			);
			// This will need to be recursive if this section ever gets more than one level deep for follow-ups
			const answersForFollowUps = followUps.map((id) => generateAnswer(id, 'No'));
			const answer = pageHeadingHasAllRelevantQuestionsConfirmed('Patient Information', {
				configuration,
				patient,
				answers: [...topLevelAnswers, ...answersForFollowUps],
			});
			expect(answer).toBe(true);
		});

		it('should return false if no answers at all ', () => {
			const answer = pageHeadingHasAllRelevantQuestionsConfirmed('Patient Information', {
				configuration,
				patient,
				answers: [],
			});
			expect(answer).toBe(false);
		});

		it('should return false if not all answered  ', () => {
			const answer = pageHeadingHasAllRelevantQuestionsConfirmed('Patient Information', {
				configuration,
				patient,
				answers: [generateAnswer('PA-59767e')],
			});
			expect(answer).toBe(false);
		});
	});

	describe(allRelevantQuestionsForPatientHaveConfirmedAnswers.name, () => {
		it('should return true if all relevant questions have confirmed answers', () => {
			const answerForEveryPossibleQuestion = answersForAllQuestionsInConfigPlusFirstIndex(configuration);
			const answer = allRelevantQuestionsForPatientHaveConfirmedAnswers({
				configuration,
				patient,
				answers: answerForEveryPossibleQuestion,
			});
			expect(answer).toBe(true);
		});

		it('should return false if not all relevant questions have confirmed answers', () => {
			const someAnswers = allIdsInPatientInformationSection.map((id) => generateAnswer(id, 'Yes'));
			const answer = allRelevantQuestionsForPatientHaveConfirmedAnswers({
				configuration,
				patient,
				answers: someAnswers,
			});
			expect(answer).toBe(false);
		});

		it('should return false if no answers', () => {
			const answer = allRelevantQuestionsForPatientHaveConfirmedAnswers({ configuration, patient, answers: [] });
			expect(answer).toBe(false);
		});
	});

	describe(questionsThatDoNotHaveConfirmedAnswers.name, () => {
		it('should return questions that do not have confirmed answers', () => {
			const toTest = 'A1005';
			const answerForEveryPossibleQuestion = answersForAllQuestionsInConfigPlusFirstIndex(configuration).filter(
				(a) => a.assessmentNumber !== toTest,
			);
			const remainingQuestions = questionsThatDoNotHaveConfirmedAnswers({
				configuration,
				patient,
				answers: answerForEveryPossibleQuestion,
			});
			expect(remainingQuestions).toEqual([toTest]);
		});

		it('should return empty array when all are answered', () => {
			const answerForEveryPossibleQuestion = answersForAllQuestionsInConfigPlusFirstIndex(configuration);
			const remainingQuestions = questionsThatDoNotHaveConfirmedAnswers({
				configuration,
				patient,
				answers: answerForEveryPossibleQuestion,
			});
			expect(remainingQuestions.length).toBe(0);
		});
	});
});

function answersForAllQuestionsInConfigPlusFirstIndex(configuration: Configuration) {
	// height and weight are in data file but Apricot Core is the one used
	const canBeSafelyIgnored = ['PA-946407', 'PA-ee1734'];
	return configuration.allQuestions
		.filter((q) => !canBeSafelyIgnored.includes(q.id))
		.flatMap((q) => [generateAnswer(q.id, 'Yes'), generateAnswer(addMultipleIndex(q.id, 1), 'Yes')]);
}

function generateAnswer(
	assessmentNumber: string,
	choice: string | string[] = 'Yes',
	confirmed = true,
): AssessmentAnswer {
	const withConfirmed = confirmed ? { checkedResponse: { choice }, generatedResponseAccepted: true } : {};
	return {
		assessmentNumber,
		generatedResponse: { choice },
		...withConfirmed,
	} as Partial<AssessmentAnswer> as AssessmentAnswer;
}
