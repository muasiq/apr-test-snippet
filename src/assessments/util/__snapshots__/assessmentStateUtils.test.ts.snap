// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assessmentStateUtils allRelevantQuestionsForPageHeading should return all relevant questions for multiple follow-up levels 1`] = `
[
  "PA-73f37a",
  "M1000",
  "PA-f2a5bc",
  "PA-15bc7b",
  "MQ3K8EDC1D",
  "PA-b08dd6",
  "O0110_A",
  "vaccinations_custom",
  "PA-105a9f",
  "MQ67LVO5CK",
  "PA-9f8d09",
  "PA-772181",
]
`;
